ffi/diplomat/src/calendar.rs

```diff
@@ -32,4 +37,56 @@ pub mod ffi {
                 .into()
         }
     }
+
+    #[diplomat::opaque]
+    #[diplomat::rust_link(icu::calendar::AnyCalendar, Enum)]
+    pub struct ICU4XCalendar(pub Arc<AnyCalendar>);
+
+    impl ICU4XCalendar {
+        /// Creates a new [`ICU4XGregorianDateTime`] from the specified date and time.
+        #[diplomat::rust_link(icu::calendar::AnyCalendar::try_new_unstable, FnInStruct)]
+        pub fn try_new(
+            provider: &ICU4XDataProvider,
+            locale: &ICU4XLocale,
+        ) -> DiplomatResult<Box<ICU4XCalendar>, ICU4XError> {
+            use icu_provider::serde::AsDeserializingBufferProvider;
+            let provider = provider.0.as_deserializing();
+
+            let kind = match (&locale.0).try_into() {
+                Ok(k) => k,
+                Err(e) => return Err(ICU4XError::from(e)).into(),
+            };
```

Comment:
Nit: This is an important section of code, and it's not clear what it is doing (selecting an extension keyword from a locale). I think I'd prefer if we didn't `impl TryFrom<Locale> for AnyCalendarKind`

Reply:
I kinda see this as analogous to the DataLocale inputs for most types; AnyCalendar directly takes an AnyCalendarKind but that's really not that different from taking in a DataLocale and reading just the calendar keyword. Basically all of our APIs do a `.try_into()` on the locale before running, we just do this with a slightly different type. AnyCalendarKind is just "the relevant subset of the datalocale", to me.

Reply:
I think the other issue here is that failing to turn a `Locale` into a `AnyCalendarKind` shoudn't fail to create an AnyCalendar; we're supposed to fall back to the locale's default. That is all code that should ideally live in Rust, not in the FFI wrapper layer or in the C++/JavaScript calling code.

I guess I would expect a function like `AnyCalendar::try_new_from_locale_unstable(&provider, &locale)`